File: A - Imperative Basics/A02 - Data Handling/A02.cs
// Author: Sady Sell Neto
// Copyright(C) 2016

using System;

static class Program {
    public static void Main() {

        // C# is a case-sensitive language. X is different from x.

        // In C#, variables are declared like in C.
        // They must be assigned right away.
        // You can use the default operator to assign the default value for
        //   a given type (usually used with generic types).
        int var = 5;
        Console.WriteLine(var);

        int dvar = default(int);
        Console.WriteLine(dvar);

        // In C#, constants are declared like in C.
        // They must be assigned right away.
        const int ten = 10;
        Console.WriteLine(ten);

        // C#'s (acutally .NET framework's) type system:
        // There are two kinds of types:
        // Value types are basic data types. Assigning one variable to another
        //   does copy the value to the other. Chaning the other has no impact
        //   on the original.
        //   Default value for these types is zero.
        // Reference types act like pointers. Assinging one variable to another
        //   will only make them reference the same region. Therefore, changiing
        //   one of them will also change the other.
        //   Default values for these types is null.
        // Both of them are objects and have methods.

        // C# basic types:

        // sbyte (System.SByte), 1-byte signed integer:
        sbyte _sbyte = 1;
        Console.WriteLine(_sbyte);

        // byte (System.Byte), 1-byte unsigned integer:
        byte _byte = 2;
        Console.WriteLine(_byte);

        // short (System.Int16), 2-byte signed integer:
        short _short = 3;
        Console.WriteLine(_short);

        // ushort (System.UInt16), 2-byte unsigned integer:
        ushort _ushort = 4;
        Console.WriteLine(_ushort);

        // int (System.Int32), 4-byte signed integer:
        int _int = 5;
        Console.WriteLine(_int);

        // uint (System.UInt32), 4-byte unsigned integer:
        uint _uint = 6u;
        Console.WriteLine(_uint);

        // long (System.Int64), 8-byte signed integer:
        long _long = 7L;
        Console.WriteLine(_long);

        // ulong (System.UInt64), 8-byte unsigned integer:
        ulong _ulong = 8uL;
        Console.WriteLine(_ulong);

        // bool (System.Boolean), 1-byte true or false value:
        bool _bool = true;
        Console.WriteLine(_bool);

        // float (System.Single), single-precision floating-point type:
        float _float = 0.9f;
        Console.WriteLine(_float);

        // double (System.Double), double-precision floating-point type:
        double _double = 0.01;
        Console.WriteLine(_double);

        // decimal (System.Decimal), floating-point type focusing on decimal
        //   part:
        decimal _decimal = 0.02m;
        Console.WriteLine(_decimal);

        // char (System.Char), unicode character:
        char _char = 'a';
        Console.WriteLine(_char);

        // string (System.String), unicode string (reference type):
        string _string = "abc";
        Console.WriteLine(_string);

        // object (System.Object), any object (reference type):
        object _object = new object();
        Console.WriteLine(_object);

        // Null reference:
        object _null = null;
        Console.WriteLine(_null);

        // Nullable types:
        // Nullable types are types that can have the null value assigned.
        // All reference types are nullable.
        // Value types can be turned into nullable appending a "?" after their
        //   type. (T? is an alias to System.Nullable<T>)
        int? _nullable = 100;
        Console.WriteLine(_nullable);

        _nullable = null;
        Console.WriteLine(_nullable);

        // C# has pointer types, but they will not be covered.

        // Type inference with the "var" keyword:
        var _inference = 200;
        Console.WriteLine(_inference);

        // Dynamic variables with the "dynamic" keyowrd (C#6):
        dynamic _dynamic = 300;
        _dynamic = "def";
        Console.WriteLine(_dynamic);

        // Type conversions:

        // C# maintains C-like cast:
        double _cast = 300.5;
        Console.WriteLine((int)_cast);

        // And introduces the "as" operator for safe casts between nullable
        //   types:
        object _as1 = new object();
        string _as2 = "ghi";

        Console.WriteLine(_as2 as object != null ? "Success" : "Failure");
        Console.WriteLine(_as1 as string != null ? "Success" : "Failure");

    }
}

File: A - Imperative Basics/A02 - Data Handling/A02.vb
' Author: Sady Sell Neto
' Copyright(C) 2016

Imports System

Module Program
    Public Sub Main()

        ' VB.NET is a case-insensitive language. X and x are the same.

        ' VB.NET variables can be unassigned or assigned right away.
        ' If unassigned, they have the same value as if assigned by C#'s
        '   default operator.
        ' This is a VB.NET variable declaration:
        Dim var As Integer = 5
        Console.WriteLine(var)

        Dim dvar As Integer
        Console.WriteLine(dvar)

        ' Constants must be assigned right away.
        ' This is a VB.NET constant declaration:
        Const ten As Integer = 10
        Console.WriteLine(ten)

        ' VB.NET's (acutally .NET framework's) type system:
        ' There are two kinds of types:
        ' Value types are basic data types. Assigning one variable to another
        '   does copy the value to the other. Chaning the other has no impact
        '   on the original.
        '   Default value for these types is zero.
        ' Reference types act like pointers. Assinging one variable to another
        '   will only make them reference the same region. Therefore, changiing
        '   one of them will also change the other.
        '   Default values for these types is Nothing.
        ' Both of them are objects and have methods.

        ' VB.NET basic types:

        ' SByte (System.SByte), 1-byte signed integer:
        Dim _sbyte As SByte = 1
        Console.WriteLine(_sbyte)

        ' Byte (System.Byte), 1-byte unsigned integer:
        Dim _byte As Byte = 2
        Console.WriteLine(_byte)

        ' Short (System.Int16), 2-byte signed integer:
        Dim _short As Short = 3
        Console.WriteLine(_short)

        ' UShort (System.UInt16), 2-byte unsigned integer:
        Dim _ushort As UShort = 4
        Console.WriteLine(_ushort)

        ' Integer (System.Int32), 4-byte signed integer:
        Dim _integer As Integer = 5
        Console.WriteLine(_integer)

        ' UInteger (System.UInt32), 4-byte unsigned integer:
        Dim _uinteger As UInteger = 6UI
        Console.WriteLine(_uinteger)

        ' Long (System.Int64), 8-byte signed integer:
        Dim _long As Long = 7L
        Console.WriteLine(_long)

        ' ULong (System.UInt64), 8-byte unsigned integer:
        Dim _ulong As ULong = 8UL
        Console.WriteLine(_ulong)

        ' Boolean (System.Boolean), 1-byte True or False value:
        Dim _boolean As Boolean = True
        Console.WriteLine(_boolean)

        ' Single (System.Single), single-precision floating-point type:
        Dim _single As Single = 0.9F
        Console.WriteLine(_single)

        ' Double (System.Double), double-precision floating-point type:
        Dim _double As Double = 0.01
        Console.WriteLine(_double)

        ' Decimal (System.Decimal), floating-point type focusing on decimal
        '   part:
        Dim _decimal As Decimal = 0.02D
        Console.WriteLine(_decimal)

        ' Char (System.Char), unicode character:
        Dim _char As Char = "a"c
        Console.WriteLine(_char)

        ' String (System.String), unicode string (reference type):
        Dim _string As String = "abc"
        Console.WriteLine(_string)

        ' Object (System.Object), any object (reference type):
        Dim _object As Object = New Object()
        Console.WriteLine(_object)

        ' Null reference:
        Dim _null As Object = Nothing
        Console.WriteLine(_null)

        ' Nullable types
        ' Nullable types are types that can have the Nothing value assigned.
        ' All reference types are nullable.
        ' Value types can be turned into nullable appending a "?" after their
        '   type or identifier. (T? Is an alias to System.Nullable(Of T))
        Dim _nullable As Integer? = 100 ' Or Dim _nullable? As Integer = 100
        Console.WriteLine(_nullable)

        _nullable = Nothing
        Console.WriteLine(_nullable)

        ' VB.NET has poor support for pointer types.

        ' One can achieve type by omitting the type:
        Dim _inference = 200

        ' One can achieve dynamic types by using "Option Strict Off".
        ' Will not be covered.

        ' Type conversions:
        ' VB.NET has function-like operators for conversions. The easiest is
        '   "CType".
        Dim _cast As Double = 300.5
        Console.WriteLine(CType(_cast, Integer))

        ' CType has a lot of aliases:
        ' CBool(x) = CType(x, Boolean)
        ' CByte(x) = CType(x, Byte)
        ' CChar(x) = CType(x, Char)
        ' CDate(x) = CType(x, Date)
        ' CDbl(x) = CType(x, Double)
        ' CDec(x) = CType(x, Decimal)
        ' CInt(x) = CType(x, Integer)
        ' CLng(x) = CType(x, Long)
        ' CObj(x) = CType(x, Object)
        ' CSByte(x) = CType(x, SByte)
        ' CShort(x) = CType(x, Short)
        ' CSng(x) = CType(x, Single)
        ' CStr(x) = CType(x, String)
        ' CUInt(x) = CType(x, UInteger)
        ' CULng(x) = CType(x, ULong)
        ' CUShort(x) = CType(x, Short)

        ' For safe conversion between nullable types, use the "TryCast"
        '   Operator, that is very smiliar to the "CType" operator.

        Dim _as1 As New Object()
        Dim _as2 As String = "ghi"

        Console.WriteLine(If(TryCast(_as2, Object) IsNot Nothing, "Success", "Failure"))
        Console.WriteLine(If(TryCast(_as1, String) IsNot Nothing, "Success", "Failure"))

    End Sub
End Module
